ai_diffusion/model.py

```diff
@@ -622,6 +633,89 @@ def _finish_job(self, job: Job, event: ClientEvent):
             self.jobs.notify_cancelled(job)
             self.progress = 0
 
+    def _apply_resize_command(self, cmd: ResizeCommand, job: Job):
+        """Record a requested canvas resize for this job.
+
+        The actual resize is applied when showing a preview or applying the result.
+        """
+        try:
+            bounds = job.params.bounds
+            job.params.bounds = Bounds(bounds.x, bounds.y, cmd.width, cmd.height)
+            job.params.metadata["resize_canvas"] = {"width": cmd.width, "height": cmd.height}
+        except Exception as e:
+            log.warning(f"Failed to store resize command from custom workflow: {e}")
+
+    def _get_resize_for_params(self, params: JobParams) -> tuple[int, int] | None:
+        resize = params.metadata.get("resize_canvas")
+        if not isinstance(resize, dict):
+            return None
+
+        try:
+            width = int(resize.get("width", 0))
+            height = int(resize.get("height", 0))
+        except Exception:
+            return None
+
+        if width <= 0 or height <= 0:
+            return None
+
+        return width, height
+
+    def _preview_canvas_size(self, params: JobParams):
+        """Resize canvas for preview, remembering original size so it can be reverted.
+
+        This only affects temporary previews. The canvas is reverted when
+        there are no history selections (see hide_preview).
+        """
+        resize = self._get_resize_for_params(params)
+        if resize is None:
+            return
+
+        width, height = resize
+        extent = self.document.extent
+        if extent.width == width and extent.height == height:
+            return
+        if self._preview_canvas_original_extent is None:
+            self._preview_canvas_original_extent = extent
+
+        try:
+            self.document.resize_canvas(width, height)
+        except Exception as e:
+            log.warning(f"Failed to resize canvas for preview from custom workflow: {e}")
+
+    def _reset_preview_canvas_size(self):
+        """Revert canvas size after preview, if it was resized just for preview."""
+        if self._preview_canvas_original_extent is None:
+            return
+
+        original = self._preview_canvas_original_extent
+        self._preview_canvas_original_extent = None
+
+        try:
+            self.document.resize_canvas(original.width, original.height)
+        except Exception as e:
+            log.warning(f"Failed to revert canvas size after preview: {e}")
+
+    def _apply_canvas_resize_for_params(self, params: JobParams):
```

Comment:
Probably can put this directly into the `apply_result` function. If `resize_canvas` is just a bool it would be something like this:
```py
if params.resize_canvas:
    self.document.resize_canvas(*image.extent)
```